python-main.js

```diff
@@ -1265,64 +1265,118 @@ function web_editor(config) {
+        $("#flashing-overlay-container").css("display", "flex");
+        $("#flashing-info").addClass('hidden');
+
         // Log error to console for feedback
-        console.log("An error occured whilst attempting to use WebUSB. Details of the error can be found below, and may be useful when trying to replicate and debug the error.");
+        console.log("An error occured whilst attempting to use WebUSB.");
+        console.log("Details of the error can be found below, and may be useful when trying to replicate and debug the error.");
         console.log(err);
+        console.trace();
 
-        // If there was an error clear dapwrapper
+        // If there was an error and quick flash is in use, then clear dapwrapper
         if(usePartialFlashing) {
-            window.dapwrapper = null;
-            window.previousDapWrapper = null;
+            if(window.dapwrapper) {
+                window.dapwrapper = null;
+            }
+
+            if(window.previousDapWrapper) {
+                window.previousDapWrapper = null;
+            }
         }
 
-        // Disconnect
+        // Disconnect from the microbit
         doDisconnect();
+       
+        var errorType;
+        var errorTitle;
+        var errorDescription;
 
-        // Error handler
-        $("#flashing-overlay-container").css("display", "flex");
-        $("#flashing-info").addClass('hidden');
+        // Determine type of error
+        switch(typeof err) {
+            case "object":
+                console.log("Caught in Promise or Error object");
+                // We might get Error objects as Promise rejection arguments
+                if (!err.message && err.promise && err.reason) {
+                    err = err.reason;
+                }
+                
+                // Determine error type
+                if (err.message === "No valid interfaces found.") {
+                    errorType = "update-req";
+                    errorTitle = err.message;
+                    errorDescription = config["translate"]["webusb"]["err"][errorType];
+                } else if (err.message === "Unable to claim interface.") {
+                    errorType = "clear-connect";
+                    errorTitle = err.message;
+                    errorDescription = config["translate"]["webusb"]["err"][errorType];
+                } else if (err.name === "device-disconnected") {
+                    errorType = "device-disconnected";
+                    errorTitle = err.message;
+                    // No additional message provided here, err.message is enough
+                    errorDescription = "";
+                } else {
+                    // Unhandled error. User will need to reconnect their micro:bit
+                    errorType = "reconnect-microbit";
+                    errorTitle = "WebUSB Error";
+                    errorDescription = config["translate"]["webusb"]["err"][errorType];
+                    if (usePartialFlashing) {
```

Comment:
Can we also add a check for the experimental flag here and in the other instances of this part of the error message?

As only the beta version of the editor can disable Quick Flash, the message "If the errors persist, try disabling Quick Flash in the beta options." should not appear in the primary editor.

Reply:
👍